clippy_lints/src/clones_into_boxed_slices.rs

```diff
@@ -0,0 +1,204 @@
+fn ty_is_slice_like(cx: &LateContext<'_>, ty: Ty<'_>) -> bool {
+    ty.is_slice()
+        || ty.is_str()
+        || ty.is_diag_item(cx, sym::cstr_type)
+        || ty.is_diag_item(cx, sym::Path)
+        || ty.is_diag_item(cx, sym::OsStr)
+}
+
+// Checks if an expression is one of the into_boxed_... methods preceded by a clone-like function
+// Then shows the lint with a suggestion that depends on the types of the inner argument and the
+// resulting Box
+impl<'tcx> LateLintPass<'tcx> for ClonesIntoBoxedSlices {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, second_method: &'tcx Expr<'_>) {
+        // Is the second method into_boxed_...?
+        if let ExprKind::MethodCall(second_method_path, first_method, _, _) = second_method.kind
+            && second_method.span.eq_ctxt(first_method.span)
+            && [
+                sym::into_boxed_str,
+                sym::into_boxed_slice,
+                sym::into_boxed_path,
+                sym::into_boxed_c_str,
+                sym::into_boxed_os_str,
+            ]
+            .contains(&second_method_path.ident.name)
+        {
+            let arg = match first_method.kind {
+                // Is the first method clone-like?
+                ExprKind::MethodCall(first_method_path, left, _, _)
+                    if [
+                        sym::to_owned,
+                        sym::clone,
+                        sym::to_string,
+                        sym::to_path_buf,
+                        sym::to_os_string,
+                        sym::to_vec,
+                    ]
+                    .contains(&first_method_path.ident.name) =>
+                {
+                    Some(left)
+                },
+                // Also check for from(...) constructor
+                ExprKind::Call(
+                    Expr {
+                        hir_id: _,
+                        kind: ExprKind::Path(QPath::TypeRelative(call_out_ty, call_path)),
+                        span: _,
+                    },
+                    args,
+                ) if call_path.ident.name == sym::from && cx.typeck_results().expr_ty(&args[0]).is_ref() => {
+                    Some(&args[0])
+                },
+                _ => None,
+            };
+
+            if let Some(arg) = arg {
+                let full_span = second_method.span.to(first_method.span);
+                let arg_ty = cx.typeck_results().expr_ty(arg);
+                let inner_ty = arg_ty.peel_refs();
+                if ty_is_slice_like(cx, inner_ty) {
+                    if second_method_path.ident.name == sym::into_boxed_path && !inner_ty.is_diag_item(cx, sym::Path) {
+                        // PathBuf's from(...) can convert from other str types,
+                        // so Path::new(...) must be used to assure resulting Box is the correct type
+                        show_lint(cx, full_span, arg, true, Some("Path::new("), "...", Some(")"));
```

Comment:
`show_lint()` should know that this suggestion should be degraded to `MaybeIncorrect` at most, because `Path` may not be have been imported.